VectorSource load events

simonseyock opened the pull request:
Resolves #9975

simonseyock commented:
This ready for review

ahocevar commented on src/ol/source/VectorEventType.js:
```diff
@@ -34,4 +34,25 @@ export default {
+
+  /**
+   * Triggered if feature loading results in an error.
+   * @event module:ol/source/TVector.VectorSourceEvent#featureloaderror
```
```suggestion
   * @event module:ol/source/Vector.VectorSourceEvent#featureloaderror
```

ahocevar commented on src/ol/featureloader.js:
```diff
@@ -130,25 +130,38 @@ export function loadFeaturesXhr(url, format, success, failure) {
+        if (success !== undefined) {
+          success();
+        }
+        sourceOrTile.addFeatures(features);
```
You'd still have to check for availabilty of `addFeatures`, because vector tiles don't have that function. But I have a better idea: Now that you have a `success` callback, this function could be improved if it wouldn't deal with adding features at all any more. Then the tile or source would take care of adding features in the `success` callback. Should be a straightforward change, if you change the `success` callback to accept `features` as argument.

simonseyock replied:
The xhr function is not actually used in the vector tiles. There exists a `defaultLoadFunction` which does this. That's why I proposed in https://github.com/openlayers/openlayers/issues/9975#issuecomment-704092256 that the functionality could be moved into a similar default function in the `source/Vector` file.

I agree that this could be done with a callback as well. I abstained from doing it, because it would break existing custom load functions. But if that's ok, I can do that.

ahocevar replied:
Let's not break existing custom load function. Good point!

Thanks for the explanations! I had missed that the function is not used by the VectorTile source any more. I don't think it's necessary to do the refactoring you proposed in https://github.com/openlayers/openlayers/issues/9975#issuecomment-704092256. I think the only thing that's left to do here is rename the `sourceOrTile` variable to `source`, and change its type accordingly.

ahocevar commented on src/ol/source/VectorEventType.js:
```diff
@@ -34,4 +34,25 @@ export default {
+   * @event module:ol/source/Vector.VectorSourceEvent#featureloadstart
+   * @api
+   */
+  FEATURELOADSTART: 'featureloadstart',
```
Maybe just a matter of taste, but I think the event names could be shorter: `loadstart`, `loadend`, `loaderror`.

simonseyock replied:
I made it this way, because it is similar to the `tileloadstart` and `imageloadstart`. But I don't see why these have to be named differently in the first place. We could also introduce these events in `source/Source` if we implemented them in `source/UTFGrid` as well. They would not be fired in `source/Tile` and `source/Source` then, but those are abstract anyways.

simonseyock replied:
I noticed that the other load events always contain the image or tile. This way the names make sense and they cannot be generalized that easily.

It could make sense to provide the loaded features in the `loadend` event for other than the default loading strategy.

ahocevar replied:
The reason why I started thinking about it was because we're loading `features`, not a `feature`. `featuresloadstart` would be a long event name, but it would be the correct name. And yes, it might make sense to provide the loaded features in the `featuresloadend` event.

ahocevar reviewed (changes requested):
Thanks for your work on this, @simonseyock! I added a few comments. Would be great if you could address them.

simonseyock commented:
I added the changes. I introduced a new property `features` on the `VectorSourceEvent`. This looks a little ugly with the other property being `feature`, I think the proper way would be to remove `feature` and just provide an array with just one element (similiar to the modify event). But again this would be a breaking change. Or should I introduce a new event type `VectorSourceLoadedEvent`?

ahocevar reviewed (commented):
No need to do further renaming of event properties or make a breaking change. I think this looks good now. I only added two suggestions for the docs.

simonseyock commented:
done.

ahocevar reviewed (approved):
Thanks, @simonseyock
